### PR TITLE
Update Config and Database routines

### PR DESCRIPTION
**Pull request by dualspiral**
After seeing your post on the Sponge Forums, I thought that I'd whip up a couple of changes that I suggest for you. I realise that they might look a little scary, but I'll do my best to explain what's going on. I realise you're a self-confessed beginner, what I'm doing is trying to show you how to write code that prevents you from repeating yourself too much - if you have to change code, you will only want to change it once!

**Config Updates**

Configurate, the system that Sponge uses to manage configuration files, is a lot more powerful than most developers realise. I'm not going to tell you about everything, but I am going to explain what I've done here, object mapping.

Before, you had a static class that initialised, set defaults and loaded the file, and then proxied through to the loader/node. The main problem you had was then you had strings everywhere, and if you changed the name of a config key in one place, you have to change them in all. This was far from ideal.

I have written you a `ConfigManager<T>` class, and a `MainConfig` data class. This uses a feature in Configurate called Object Mapping. Using this, you can define your configuration as a class in Java, if you look at the `MainConfig` class below, (or [if you look here](https://github.com/dualspiral/Guilds/blob/b2c4bc148cee246e0a2ffff29a5564afd478ec80/src/main/java/com/meronat/guilds/configuration/MainConfig.java)) you'll see that I've created an object that mirrors your Config file and has defaults set. Each field that will contain a configuration value has a `@Setting` annotation with the name of the config variable, and also the ability to add a comment. Note that the `MainConfig` file also has the `@ConfigSerializable` annotation, this tells Configurate that this object can be used in configuration.

If you want sections, you can add a `@Setting` field that contains another class with the `@ConfigSerializable` annotation, like I do in Nucleus [here in the base Config class](https://github.com/dualspiral/Guilds/blob/b2c4bc148cee246e0a2ffff29a5564afd478ec80/src/main/java/com/meronat/guilds/configuration/MainConfig.java) with this [class as a subsection](https://github.com/NucleusPowered/Nucleus/blob/master/src/main/java/io/github/nucleuspowered/nucleus/modules/core/config/KickOnStopConfig.java).

I've setup Guilds to then load this `MainConfig` file, and save/create it when you start the server, to add any defaults you put into this file. To set a default, just set the value of the field when the class is created.

Guilds will reload the file when anyone runs `/sponge plugins reload`.

If you were to create another config file, you should be able to just use the `ConfigManager<>` class using a different `@ConfigSerializable` class, just provide a different type token, loader and a supplier that creates a new object and it should just work.

**Storage class**

The biggest problem you had here is that your class was massive, and full of repeated code - you should try to follow the DRY principle - [don't repeat yourself](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)! You were repeating the connection opening and closing code, which would be _really_ tedious to update everywhere. So, I've shortened it for you, and put all the database connection stuff in one place.

First: I used Java's `try-with-resources` when opening a connection. By opening the connection in the `try` statement, it will automatically be closed when you leave the `try` block, either by exception, or otherwise. It's the same as having a `finally` block - but you don't need to write it.

Second: I have used some Java 8 syntax by means of `Consumers`, `Suppliers` and lambdas. I noticed that as you have three types of operation, write single row, write multiple/batch of rows and read rows, I took that logic, and wrote it so that you just need to provide a SQL statement and a function that updates your prepared statements and/or reads your results, and a supplier that gives you an error message if things don't work. That way, you don't have to copy and paste connection logic each time.

[An example of updating a row is here](https://github.com/dualspiral/Guilds/blob/b2c4bc148cee246e0a2ffff29a5564afd478ec80/src/main/java/com/meronat/guilds/configuration/Storage.java#L298-L302). I provide the SQL statement, how the variables are added to the PreparedStatement, and a supplier that gives the error message if something throws an exception. If you have to add other SQL operations in the future, hopefully these methods will make writing the CRUD operations a little bit faster because you don't have to copy/paste all the `try/catch/finally` stuff.

I also updated your Toggles that are stored in the DB to set and get their names, rather than use `toString()`, that way, you can be sure that `Enum.valueOf()` gets what you're expecting - normally, you should use `toString()` as the JavaDocs suggest, but you need the exactness in this case to go back and forth, and if you override `toString()` in the future, you won't break the plugin.

As an aside, I saw there was some discussion about using Sponge Data to store some of this information. It's a option, and one that I would still consider, but I must impress on your that _you do not have to do it that way if you don't want to_, just like you are free to not take parts/all of this PR. Sponge Data is a very nice feature of Sponge for attaching data to blocks, players, locations etc., but it is not a silver bullet for everything. SQL might be better for you here, especially if you start running things async - but I do not know this and it's something to research. Do what _you_ think is best for _your_ plugin.

**Economy Service**

I added an event to listen for Economy Service changes, which includes a consumer that references the `private static` field in the main plugin class so that you do not have to make the access public. I've also set it up to _always_ get the service, and just not return it if economy support is disabled.

**Error on loading**

There isn't a way to disable a plugin that errors through the lifecycle, so the best thing to do is just use a boolean flag if you encounter an error and return from other loading events if that flag is true. I've done that for you.

I hope that this PR is useful to you, and that you learn something from it! If you have questions, feel free to come back at me!

### CODE REVIEW

**Comment by Meronat** [comment]
Wow @dualspiral, thank you very much for this. I wouldn't have thought of most of this, and it's so nice to see my database class short again. It did seem poor to have every method so long with all of those catch and finally statements, so this is a refreshing and smart way to do it.

The use of lambdas, such as when setting the default permission actually makes sense when translating it from what I had to that. They're a bit complicated still, but after viewing this I feel like I may be able to write some more on my own. I do have some questions though, I'm not sure I completely understand what the Consumers and Suppliers are doing and how they differentiate, would you mind elaborating? 

This use of Configurate here is actually very useful as well and makes a lot of sense, I love that you can simply return the values you have as Settings. Your use of the Math min and max is much simpler than what I was doing too and you account for the SQL max length which is a case I did not think of. 

Oh, also on plugin reload do I have to worry about anything related to my database, such as reloading any of the data stored there or will everything I have in memory stay? 

There's a lot for me to learn from these changes, especially in my storage class, and I'm excited to do so and hopefully even make use of some of these practices in my own code in the future. Thank you for the assistance, all of these changes are great for the plugin and my learning, I appreciate it. I hope to merge this later today :)

**Comment by dualspiral** [comment]
>  I'm not sure I completely understand what the Consumers and Suppliers are doing and how they differentiate

Quick recap of what they are in general:

A _Consumer_ is a function that takes an argument and uses it, and returns nothing. It "consumes" it.
A _Supplier_ is a function that takes no arguments, and provides (or supplies) an object.

I realise I should have commented the functions. [I've done that now](https://github.com/Meronat/Guilds/pull/1/commits/224ea8f509044a3f182a87610a55392b99d91c70), but I'll explain what a couple of them do here:

---

```java
private void executeQuery(String sql, @Nullable ThrownConsumer<PreparedStatement, SQLException> preparedStatementConsumer, ThrownConsumer<ResultSet, SQLException> resultSetConsumer, Supplier<String> errorMessageSupplier)
```

`executeQuery` is the function that does SELECT operations.

**`ThrownConsumer<PreparedStatement, SQLException> preparedStatementConsumer`** takes the `PreparedStatement` from the connection, and has the SQL inserted into it. In the consumer, you set the variables within it, `ps.setString(1, "String")`, for example. It's not executed in there, all this method does is to morph the statement.

**`ThrownConsumer<ResultSet, SQLException> resultSetConsumer`** has the `ResultSet` from the query that you can do your data operations on.

**`Supplier<String> errorMessageSupplier`** provides the error message to use if the database operations fail.

---

```java
private void executeSingle(String sql, ThrownConsumer<PreparedStatement, SQLException>preparedStatementConsumer, Supplier<String> errorMessageSupplier)
```

**`ThrownConsumer<PreparedStatement, SQLException> preparedStatementConsumer`** and **` Supplier<String> errorMessageSupplier`** are the same as above, but this is intended for doing inserts and updates.

**Comment by dualspiral** [comment]
> Oh, also on plugin reload do I have to worry about anything related to my database, such as reloading any of the data stored there or will everything I have in memory stay? 

No, I have just written it to reload the configuration only, your database and in memory data will remain untouched.

File: src/main/java/com/meronat/guilds/Guilds.java [hunk 1]
```diff
@@ -185,13 +172,19 @@ private void registerListeners() {
         spongeEventManager.registerListeners(this, new DeathListener());
         spongeEventManager.registerListeners(this, new InteractListener());
         spongeEventManager.registerListeners(this, new SpawnListener());
+        spongeEventManager.registerListeners(this, new ReloadEventListener());
+
+        // Tell the ChangeServiceListener how to update the economy service.
+        spongeEventManager.registerListeners(this, new ChangeServiceListener(newService -> Guilds.economyService = newService));
```
Comment (dualspiral) [inline src/main/java/com/meronat/guilds/Guilds.java]: I'll quickly explain _this_ lambda. This is a `Consumer<EconomyService>`. What it does it tells the `ChangeServiceListener` how to update the `private static` variable in this class, without actually making the variable public. That's the power of this lambda, I make the assignment in this class, but execute it in another without resorting to reflection (which I won't go into here), letting me keep the variable private!

**Comment by Meronat** [comment]
Okay, the Consumers, Lambdas, and Suppliers all seem very useful actually, but they're a different way of thinking/coding that I will need to adjust to. Thank you again for all the help, I can't wait to continue extending this plugin with this pull request and some of what I learned from it.

**Comment by dualspiral** [comment]
You're welcome!

**Pull request merged**